test/e2e/operator_test.go

```diff
@@ -2585,6 +2585,93 @@ func TestCustomErrorpages(t *testing.T) {
 	}
 }
 
+// This test validates BZ2054200: Don't delete services that are not directly owned by this controller.
```

Comment:
We usually do a unit test for bug fixes, and e2e tests for new features.  It doesn't hurt to have an e2e test, but a unit test should be done.

Reply:
I agree - but I don't think we can do a unit test here, right? I'm curious, can you unit test a function that makes an API (see r.client.Get...)? I think that requires a cluster.

Reply:
You can mock out two services on the ic, one that is owned and one that is not.

Reply:
So here's the function I made changes to and we'd want to unit test:

```
func (r *reconciler) currentNodePortService(ic *operatorv1.IngressController) (bool, *corev1.Service, error) {
	service := &corev1.Service{}
	if err := r.client.Get(context.TODO(), controller.NodePortServiceName(ic), service); err != nil {
		if errors.IsNotFound(err) {
			return false, nil, nil
		}
		return false, nil, err
	}

	// BZ2054200: Don't delete services that are not directly owned by this controller
	if service.Labels[manifests.OwningIngressControllerLabel] != ic.Name {
		return false, nil, nil
	}

	return true, service, nil
}

```
I don't see a way to mock out the two services when the function itself queries the API for the services. It would fail every time in a unit test because there is no API in a unit test right?

Reply:
There's probably more complicated ways to do it, but if you extract the code you added into its own function, you can test it more easily by feeding it a mock service.  You'll want to make this iterate over of list of service, but you get the idea:

```
func checkIfServiceOwned(service corev1.Service, owner string)bool {
        if service.Labels[manifests.OwningIngressControllerLabel] != owner {
		return false
	}
        return true
}
```

Reply:
Oh I see now. Yes - I will address this in my next update since I made it a separate function like you said. thanks!